homeassistant/components/insteon/utils.py

```diff
@@ -47,84 +64,104 @@
-        state_name = device.states[group].name
-        button = (
-            "" if state_name == BUTTON_PRESSED_STATE_NAME else state_name[-1].lower()
-        )
-        schema = {CONF_ADDRESS: address.hex}
-        if button:
-            schema[EVENT_CONF_BUTTON] = button
-        event = EVENT_BUTTON_ON if val else EVENT_BUTTON_OFF
-        _LOGGER.debug(
-            "Firing event %s with address %s and button %s", event, address.hex, button
-        )
-        hass.bus.fire(event, schema)
+        if button and button[-2] == "_":
+            button_id = button[-1].lower()
+        else:
+            button_id = None
+
+        schema = {CONF_ADDRESS: address}
+        if button_id:
+            schema[EVENT_CONF_BUTTON] = button_id
+        if name == ON_EVENT:
+            event = EVENT_GROUP_ON
+        if name == OFF_EVENT:
+            event = EVENT_GROUP_OFF
+        if name == ON_FAST_EVENT:
+            event = EVENT_GROUP_ON_FAST
+        if name == OFF_FAST_EVENT:
+            event = EVENT_GROUP_OFF_FAST
+        _LOGGER.debug("Firing event %s with %s", event, schema)
+        hass.bus.async_fire(event, schema)
+
+    for group in device.events:
+        if isinstance(group, int):
+            for event in device.events[group]:
+                if event in [
+                    OFF_EVENT,
+                    ON_EVENT,
+                    OFF_FAST_EVENT,
+                    ON_FAST_EVENT,
+                ]:
+                    _LOGGER.debug(
+                        "Registering on/off event for %s %d %s",
+                        str(device.address),
+                        group,
+                        event,
+                    )
+                    device.events[group][event].subscribe(
+                        async_fire_group_on_off_event, force_strong_ref=True
+                    )
+
+
+def register_new_device_callback(hass, config):
+    """Register callback for new Insteon device."""
 
     @callback
-    def async_new_insteon_device(device):
+    def async_new_insteon_device(address=None):
         """Detect device from transport to be delegated to platform."""
-        ipdb = IPDB()
-        for state_key in device.states:
-            platform_info = ipdb[device.states[state_key]]
-            if platform_info and platform_info.platform:
-                platform = platform_info.platform
-
-                if platform == "on_off_events":
-                    device.states[state_key].register_updates(_fire_button_on_off_event)
-
-                else:
-                    _LOGGER.info(
-                        "New INSTEON device: %s (%s) %s",
-                        device.address,
-                        device.states[state_key].name,
-                        platform,
-                    )
+        _LOGGER.debug(
+            "Adding new INSTEON device to Home Assistant with address %s", address
+        )
+        hass.async_add_executor_job(devices.async_save, hass.config.config_dir)
```

Comment:
This should preferably be awaited, possibly with an async lock wrapping it if the device callbacks can come rapidly. Otherwise we could saturate the executor thread pool.